pytential/qbx/distributed.py

```diff
@@ -0,0 +1,1071 @@
+
+    def get_local_fmm_expansion_wrangler_extra_kwargs(
+            self, actx, src_idx, target_kernels, tree_user_source_ids, arguments,
+            evaluator):
+        mpi_rank = self.comm.Get_rank()
+
+        kernel_extra_kwargs = {}
+        source_extra_kwargs = {}
+
+        if mpi_rank == 0:
+            kernel_extra_kwargs, source_extra_kwargs = \
+                self.get_fmm_expansion_wrangler_extra_kwargs(
+                    actx, target_kernels, tree_user_source_ids,
+                    arguments, evaluator)
+
+        # kernel_extra_kwargs contains information like helmholtz k, which should be
+        # picklable and cheap to broadcast
+        kernel_extra_kwargs = self.comm.bcast(kernel_extra_kwargs, root=0)
+
+        # Broadcast the keys in `source_extra_kwargs` to worker ranks
+        source_arg_names = None
+        if mpi_rank == 0:
+            source_arg_names = list(source_extra_kwargs.keys())
+        source_arg_names = self.comm.bcast(source_arg_names, root=0)
+
+        for arg_name in source_arg_names:
+            if arg_name != "dsource_vec":
+                raise NotImplementedError
+
+            # Broadcast the global source array to worker ranks
+            global_array_host = None
+            if mpi_rank == 0:
+                global_array_host = actx.to_numpy(source_extra_kwargs[arg_name])
+            global_array_host = self.comm.bcast(global_array_host, root=0)
+
+            # Compute the local source array independently on each worker rank
+            local_array_host = np.empty_like(global_array_host)
+            for idim, global_array_idim in enumerate(global_array_host):
+                local_array_host[idim] = global_array_idim[src_idx]
+
+            source_extra_kwargs[arg_name] = actx.from_numpy(local_array_host)
+
+        return kernel_extra_kwargs, source_extra_kwargs
+
+    def exec_compute_potential_insn(self, actx, insn, bound_expr, evaluate,
+            return_timing_data):
+        extra_args = {}
+
+        # Broadcast whether to use direct evaluation or FMM
+        use_direct = True
+        if self.comm.Get_rank() == 0:
+            use_direct = self.fmm_level_to_order is False
+        use_direct = self.comm.bcast(use_direct, root=0)
+
+        if use_direct:
+            func = self.exec_compute_potential_insn_direct
+            extra_args["return_timing_data"] = return_timing_data
+        else:
+            func = self.exec_compute_potential_insn_fmm
+            extra_args["fmm_driver"] = None
+
+        if self.comm.Get_rank() == 0:
+            return self._dispatch_compute_potential_insn(
+                    actx, insn, bound_expr, evaluate, func, extra_args)
+        else:
+            return func(actx, insn, bound_expr, evaluate, **extra_args)
+
+    def exec_compute_potential_insn_direct(self, *args, **kwargs):
+        if self.comm.Get_rank() == 0:
+            return super().exec_compute_potential_insn_direct(*args, **kwargs)
```

Comment:
```suggestion
            # FIXME Executes single-rank
            return super().exec_compute_potential_insn_direct(*args, **kwargs)
```
(not for this PR) :slightly_smiling_face:

Reply:
Add a warning here.

Reply:
Addressed in https://github.com/inducer/pytential/pull/224/commits/1a384f43726e941ada86d4ffc23d1a18b1b5aef4.